h2o-algos/src/main/java/hex/glm/GLM.java

```diff
@@ -1634,73 +1725,158 @@ public GramSolver(Gram gram, double[] xy, boolean intercept, double l2pen, doubl
       _gram = gram;
       // Try to pick optimal rho constant here used in ADMM solver.
       //
-      // Rho defines the strength of proximal-penalty and also the strentg of L1 penalty aplpied in each step.
-      // Picking good rho constant is tricky and greatly influences the speed of convergence and precision with which we are able to solve the problem.
+      // Rho defines the strength of proximal-penalty and also the strength of L1 penalty applied in each step.
+      // Picking good rho constant is tricky and greatly influences the speed of convergence and precision with 
+      // which we are able to solve the problem.
       //
       // Intuitively, we want the proximal l2-penalty ~ l1 penalty (l1 pen = lambda/rho, where lambda is the l1 penalty applied to the problem)
       // Here we compute the rho for each coordinate by using equation for computing coefficient for single coordinate and then making the two penalties equal.
       //
-      int ii = intercept ? 1 : 0;
-      int icptCol = gram.fullN()-1;
+      int ii = intercept ? 1 : 0; // if intercept is absent, mess with it and won't care
+      int classIndBound = gram._multinomialSpeedUp ? nclass : 1;
+      int coeffPClass = activeColAll == null?xy.length / classIndBound:-1;
+      int icptCol = gram._multinomialSpeedUp?(activeColAll==null?(coeffPClass-1):-1):(gram.fullN()-1);
       double[] rhos = MemoryManager.malloc8d(xy.length);
-      double min = Double.POSITIVE_INFINITY;
-      for (int i = 0; i < xy.length - ii; ++i) {
-        double d = xy[i];
-        d = d >= 0 ? d : -d;
-        if (d < min && d != 0) min = d;
-      }
-      double ybar = xy[icptCol];
-      for (int i = 0; i < rhos.length - ii; ++i) {
-        double y = xy[i];
-        if (y == 0) y = min;
-        double xbar = gram.get(icptCol, i);
-        double x = ((y - ybar * xbar) / ((gram.get(i, i) - xbar * xbar) + l2pen));///gram.get(i,i);
-        rhos[i] = ADMM.L1Solver.estimateRho(x, l1pen, lb == null ? Double.NEGATIVE_INFINITY : lb[i], ub == null ? Double.POSITIVE_INFINITY : ub[i]);
-      }
-      // do the intercept separate as l1pen does not apply to it
-      if (intercept && (lb != null && !Double.isInfinite(lb[icptCol]) || ub != null && !Double.isInfinite(ub[icptCol]))) {
-        int icpt = xy.length - 1;
-        rhos[icpt] = 1;//(xy[icpt] >= 0 ? xy[icpt] : -xy[icpt]);
+
+      int lastIndex = coeffPClass - ii;
+      if (activeColAll==null) { // find rhos for full length beta
```

Comment:
this method is far too long, the comment is indicative of this being a good place to split it